src/Symfony/Component/Messenger/Middleware/HandlesRecordedMessagesMiddleware.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Middleware;
+
+use Symfony\Component\Messenger\Exception\MessageRecorderHandlerException;
+use Symfony\Component\Messenger\MessageBusInterface;
+use Symfony\Component\Messenger\MessageRecorderInterface;
+
+/**
+ * A middleware that takes all recorded messages and dispatch them to the bus.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ * @author Matthias Noback <matthiasnoback@gmail.com>
+ */
+class HandlesRecordedMessagesMiddleware implements MiddlewareInterface
+{
+    private $messageRecorder;
+    private $messageBus;
+
+    public function __construct(MessageBusInterface $messageBus, MessageRecorderInterface $messageRecorder)
+    {
+        $this->messageRecorder = $messageRecorder;
+        $this->messageBus = $messageBus;
+    }
+
+    public function handle($message, callable $next)
+    {
+        // Make sure the recorder is empty before we begin
+        $this->messageRecorder->erase();
+
+        try {
+            $next($message);
+        } catch (\Throwable $exception) {
+            $this->messageRecorder->erase();
+
+            throw $exception;
+        }
+
+        $exceptions = array();
+        while (!empty($recordedMessages = $this->messageRecorder->fetch())) {
+            $this->messageRecorder->erase();
```

Comment:
If the use-case is always `fetch` and `erase`... I'm pretty sure we can find better than `fetch` + `erase/reset`.

Reply:
Im not sure what you mean here.